proposals/gateway-plusplus-phase1.md

```diff
@@ -0,0 +1,93 @@
+
+We should lean into this rather than push against it. In the short term, we can’t expect every user and developer to have a full implementation of the protocol. Strengthening the Gateway protocol to meet user needs is the path of least resistance and can enable numerous thin clients to be built for IPFS that can run in virtually any environment.
+
+This proposal establishes "Gateway++" as a meta-project for meeting these user needs with extensions to our gateway protocols and infra.
+
+## Problem Statement
+
+For reading data, the IPFS Gateway is already serving these users quite well. Not only does it allow them to read data from the IPFS network without running a full node, they are able to integrate with existing HTTP caching infrastructure to improve performance.
+
+However, the user needs for writing data to the network are not being met well by our current solutions. We have two interfaces that support writes over HTTP and neither was designed as a full solution to these user needs.
+
+- Pinning Service API.
+    - 😡 Requires you to already be running a full IPFS node as a client.
+    - 😡 Is not transactional. You have to poll in order to tell when the data is actually available.
+    - 😃 Is intended to be multi-tenant and authenticated using a flexible bearer token.
+- HTTP RPC API
+    - 😡 Is single-tenant. No authentication system.
+        - The RPC API is designed for a single user to manage a remote node, it was not designed and built to be multi-tenant and would look very different if it were.
+    - 😡 Is not always transactional. Many write calls return before they are available.
+
+These interfaces were built with particular user profiles in mind. The users we’re seeing in NFTs (including our own service as a user of IPFS) are not well served by these interfaces and we need a new “project” definition (Gateway++) that is centered around these user needs.
+
+Developers are building applications that have to run in a lot of constrained environments that can’t support a full node. Strengthening the Gateway to serve these needs over HTTP leverages endpoints they are already using and provides the features they need in a stateless protocol that has ubiquitous support.
+
+We are never going to hear “Sorry, I can’t use HTTP in this environment” from a developer.
+
+## Phase 1
+
+We should expect this to be a long term project with additional proposals in due time.
+
+For now, we have a list of high priority tasks that need to be fixed.
+
+In [nft.storage](http://nft.storage) we have the following high priority needs:
+
+- Add the Pinning API to ipfs-cluster.
+- Add transactional CAR file uploads to the Pinning API.
+
```

Comment:
It would be great to have https://github.com/ipfs/go-ipfs/issues/6129 as an addition to the HTTP API. It would allow users not using an IPFS node to assess the correct behaviour of the gateway(++).